docs/proposals/service-design.md

```diff
@@ -0,0 +1,86 @@
+The service discovery, commuunication should be considered in this feature.
+Developers can define services on top of the ddeployments deployed on the cloud or edges in a unified manner without caring about their location. The service name is the only thing they need to know to do the communication. For example, the following address can be used for a REST  request:
+`http://{service_name}/resources`
+
+## Motivation
+The cloud native and microservice architecture becoming more and more popular and the edge node is becoming more and more powerful. The user want to separate their application to multiple microservice and some of them need to be deployed in edge node to get close to the data generated by deivces. By simply the development, the user need to use the same michanism to do the service discovery and communication with what they are doing on cloud.
+
+## Constraints and Assumptions
+1. Since the edge environment can be some non-standard system to running on some specific devices so we should not rely on the kernel network capability. In that way, we need to use userspace proxy to implement the data proxy on edge.
+2. With the same reason, clusterIP is not supported on Edge.
+3. With the same reason, no DNS services available for edge.
+4. The identification and authentication should be handled by microservices.
+5. When service communication between edge to edge, they may work offline. With the purpose to deliver a "expected result" to the user system, the service definition should be push down to the edge right after the definition operation. That is to say, the istio model may not be suitable for edge, in istio model, the service definition is pull from K8S master when the service calling happens.
+6. Only support HTTP communication in the first step
+
+## Use cases
+<img src="../images/proposals/service-design.png">
+
+### Register the service located on cloud
+* CS0.1: User create service(type ClusterIP) in K8S API Server
+* CS0.2: EC get the service definition and create router rules with the edge servicebus as source and service clusterIP as target.
+* CS0.3: Send the service definition to edge
+
+### Register the service located on edge
+* CS1: User create service(headless service without ClusterIP) in K8S API Server
+* CS2, CS3: EC get the service definiton and update the service ClusterIP to itself. Then create router rule with the EC(router) address as the source and edge servicebus as target.
+* CS4: Send the service definition to edge
+
+### Service discovery on cloud
+* Application access service with the same mechanism no matter it located on cloud or edge
+
+### Service discovery on edge
+* Need to setup http_proxy in the container(can be done automatically) with the address of ServiceBus. Then the application can vist the service provider via `http://{service_name}/resource`. ServiceBus get the service name part from the request as a reverse proxy, then pick up one POD to redirect the request.
```

Comment:
can be done automatically -> this needs to added as an environment variable by edged when the container comes up ?
since servicebus will a beehive module which doesn't expose any endpoint , how will the address of the servicebus look like ?

Reply:
there will be a environment variable as well. The servicebus itself listensing the docker0 address or 127.0.0.1 to those containers using host network.